### PR TITLE
Clean up implementation of checked integer arithmetic

### PR DESCRIPTION
**Pull request by eschnett**
- Move checked_\* implementation in Base into a module Checked in a new file checked.jl
- Provide missing checked_\* functions for BigInt (BigInt always checks)
- Provide missing type promotions for checked_\* functions
- Move all checked_\* tests into their on file test/checked.jl
- Clean up handling LLVM codegen bugs
- Introduce new function jl_get_LLVM_VERSION which works already early during bootstrap
- Rename intrinsics to have an "_int" suffix, like all other integer intrinsics
- In coreimg.jl, emulate checked arithmetic by regular (unchecked) arithmetic

### CODE REVIEW

File: test/checked.jl [hunk 1]
```diff
@@ -0,0 +1,224 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+# Checked integer arithmetic
```
Comment (tkelman) [inline test/checked.jl]: needs to be added to choosetests.jl or it won't run

File: src/sys.c [hunk 2]
```diff
@@ -724,6 +726,15 @@ JL_DLLEXPORT void jl_raise_debugger(void)
 #endif // _OS_WINDOWS_
 }
 
+JL_DLLEXPORT uint32_t jl_get_LLVM_VERSION(void)
```
Comment (tkelman) [inline src/sys.c]: I don't think we want this here, now that libjulia can be built with a stub no-LLVM version of codegen

Comment (eschnett) [inline src/sys.c]: Can you suggest an alternative? Would `intrinsics.cpp` be a good place?

**Comment by eschnett** [comment]
Tests are now running, moved LLVM-specific function to `intrinsics.cpp`.

File: src/intrinsics.cpp [hunk 3]
```diff
@@ -35,6 +35,18 @@ static void jl_init_intrinsic_functions_codegen(Module *m)
 #undef ALIAS
 }
 
+extern "C" {
+  JL_DLLEXPORT uint32_t jl_get_LLVM_VERSION(void);
```
Comment (tkelman) [inline src/intrinsics.cpp]: don't really need the braces for a single function

4 space indent in the function body

**Comment by eschnett** [comment]
Corrected indentation.

File: base/checked.jl [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    end
+end
+# Use these definitions to test the non-LLVM implementations
+# typealias BrokenSignedInt SignedInt
+# typealias BrokenUnsignedInt UnsignedInt
+# typealias BrokenSignedIntMul SignedInt
+# typealias BrokenUnsignedIntMul UnsignedInt
+
+# A tight `widen`
+nextwide(::Type{Int8}) = Int16
```
Comment (simonbyrne) [inline base/checked.jl]: What's the advantage of `nextwide` over `widen`?

Comment (ScottPJones) [inline base/checked.jl]: `widen` jumps right away to `Int64`, which, if you are going to be allocating an array, would be very wasteful.

Comment (simonbyrne) [inline base/checked.jl]: But why is it useful in this case?

File: base/checked.jl [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+# These unions are used for checked_{mul,div,rem}:
+#     BrokenSignedIntMul
+#     BrokenUnsignedIntMul
+
+# This code runs early during bootstrap, and we can't use Julia's version
+# strings yet
+const llvm_version = Int(ccall(:jl_get_LLVM_VERSION, UInt32, ()))
+if llvm_version >= 30500
+    if WORD_SIZE == 32
+        typealias BrokenSignedInt Union{}
```
Comment (ScottPJones) [inline base/checked.jl]: Why do you need `BrokenSignedInt`/`BrokenUnsignedInt`? They are the same (nothing) for all possible cases here.

Comment (eschnett) [inline base/checked.jl]: See the comment just above. They are the same at the moment, but as we explore other combinations of system architectures and LLVM versions, we may need to tag some types as broken. I started out this way, and then pruned the set of broken types and LLVM combinations. I'm not quite sure yet that I didn't over-prune.

File: base/checked.jl (hunk 4 above)
Comment (eschnett) [inline base/checked.jl]: It's useful for two reasons:
- 32-bit arithmetic may be cheaper than 64-bit arithmetic, in particular for multiplication and division.
- If (my dream!) the compiler auto-vectorizes the code, then smaller types allow packing more operations into a vector. 64-bit integer arithmetic is very difficult to vectorize on x86, 32-bit integer arithmetic is much easier.

Comment (simonbyrne) [inline base/checked.jl]: I see your point, but if there are genuine gains to be had (some benchmarks would be useful), then we should modify `widen` and `widemul` more generally. Just having it here is somewhat odd.

Comment (eschnett) [inline base/checked.jl]: See <http://www.agner.org/optimize/instruction_tables.pdf> for raw benchmarks for various instructions on Intel architecture. I'm looking e.g. at page 188 for the Haswell microarchitecture. There we find for `idiv` (signed integer division):
- 32 bit: 22-29 cycles latency, 8-11 cycles inverse throughput (i.e. cycles/instruction)
- 64 bit: 39-103 cycles, 24-81 cycles inverse throughput
I've ran a "real" benchmark as well, and can roughly confirm these numbers. On my system, I find
- 32 bit: about 8 cycles/instruction
- 64 bit: about 26 cycles/instruction

Should we change the definition of `widen`, or should we introduce `nextwide` into Base? Should `widen` be system-dependent, or only distinguish between 32-bit and 64-bit word sizes?

According to Agner's table, there is no real benefit to go below 32 bits on a modern 64-bit Intel architecture.

Comment (simonbyrne) [inline base/checked.jl]: I think this is a good argument to change the behaviour of `widen` (I don't think we want 2 functions which do basically the same thing), but I think that should be a separate pull request.

(In this case, the widening is only used for multiplication, and the performance stats are the same for 32/64bit).

File: base/checked.jl [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+function checked_mul{T<:SignedInt}(x::T, y::T)
+    box(T, checked_smul_int(unbox(T,x), unbox(T,y)))
+end
+function checked_mul{T<:UnsignedInt}(x::T, y::T)
+    box(T, checked_umul_int(unbox(T,x), unbox(T,y)))
+end
+function checked_mul{T<:BrokenSignedIntMul}(x::T, y::T)
+    r = nextwide(x) * nextwide(y)
+    !(typemin(T) <= r <= typemax(T)) && throw(OverflowError())
+    T(r)
```
Comment (simonbyrne) [inline base/checked.jl]: It might be faster here to use `r % T`, as that doesn't do checking (it's also possible that LLVM is able to optimise the branch out).

**Comment by eschnett** [comment]
I'm now using `widen`, and also `r %T` for performance.

File: base/checked.jl [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+function checked_mul end
+
+function checked_mul{T<:SignedInt}(x::T, y::T)
+    box(T, checked_smul_int(unbox(T,x), unbox(T,y)))
+end
+function checked_mul{T<:UnsignedInt}(x::T, y::T)
+    box(T, checked_umul_int(unbox(T,x), unbox(T,y)))
+end
+function checked_mul{T<:BrokenSignedIntMul}(x::T, y::T)
+    r = widen(x) * widen(y)
```
Comment (simonbyrne) [inline base/checked.jl]: I don't mean to be pedantic, but you can use `widemul` for this.

File: base/checked.jl (hunk 5 above)
Comment (ScottPJones) [inline base/checked.jl]: I should have been clearer - I'm not saying that you shouldn't have the 4 cases, but I think the construction could be cleaner and more compact. For example, the following is only 15 lines:
```julia
typealias BrokenSignedInt Union{}
typealias BrokenUnsignedInt Union{}
let signedintmul = [Int128]
    unsignedintmul = [UInt128]
    if WORD_SIZE == 32
        push!(signedintmul, Int64)
        push!(unsignedintmul, UInt64)
    end
    if llvm_version < 30500
        push!(signedintmul, Int8)
        push!(unsignedintmul, UInt8)
    end
    @eval typealias BrokenSignedIntMul $(Union{signedintmul...})
    @eval typealias BrokenUnsignedIntMul $(Union{unsignedintmul...})
end
```

Comment (eschnett) [inline base/checked.jl]: I tried to separate the cases out for testing, so that if I found a new broken case, I could add it. I didn't know ahead of time that e.g. `Int64` is broken for all 32-bit architectures, and that `Int8` is broken just for LLVM <3.5. Will simplify.

Comment (eschnett) [inline base/checked.jl]: @ScottPJones This code runs too early during bootstrap, and array operations are not available yet. But one can write it purely with type unions.

Comment (ScottPJones) [inline base/checked.jl]: Ah, OK.  That's bitten me also several times when trying to fix things with the Unicode support.

Comment (ScottPJones) [inline base/checked.jl]: Your approach with type unions is nicer than mine anyway! :grinning:

**Comment by jakebolewski** [comment]
LGTM.  @tkelman the "checked" tests are included in `choosetests.jl`.  Is this waiting on #14414?

**Comment by eschnett** [comment]
@jakebolewski No, this is independent.

**Comment by simonbyrne** [comment]
LGTM as well.

**Comment by eschnett** [comment]
Rebased.

File: base/int.jl [hunk 8]
```diff
@@ -412,13 +375,48 @@ if WORD_SIZE == 32
         (lolo&0xffffffffffffffff) + UInt128(w1)<<64
     end
 
-    div(x::Int128, y::Int128) = Int128(div(BigInt(x),BigInt(y)))
-    div(x::UInt128, y::UInt128) = UInt128(div(BigInt(x),BigInt(y)))
+    function div(x::Int128, y::Int128)
+        try
```
Comment (JeffBezanson) [inline base/int.jl]: Using try/catch here is going to be really slow. Would be better to use a test and branch for the one invalid case.

Comment (eschnett) [inline base/int.jl]: See #14436.
